voting/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_sdk::collections::Map;
+use near_sdk::{env, near_bindgen, AccountId, Balance, EpochHeight};
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+type Timestamp = u64;
+
+/// Voting contract for unlocking transfers. Once the majority of the stake holders agree to
+/// unlock transfer, the time will be recorded and the voting ends.
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct VotingContract {
+    /// How much each validator votes
+    votes: Map<AccountId, Balance>,
+    /// Total voted balance so far.
+    total_voted_stake: Balance,
+    /// When the voting ended. `None` means the poll is still open.
+    result: Option<Timestamp>,
+    /// Epoch height when the contract is touched last time.
+    last_epoch_height: EpochHeight,
+}
+
+impl Default for VotingContract {
+    fn default() -> Self {
+        env::panic(b"Voting contract should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl VotingContract {
+    #[init]
+    pub fn new() -> Self {
+        VotingContract {
+            votes: Map::new(b"a".to_vec()),
+            total_voted_stake: 0,
+            result: None,
+            last_epoch_height: 0,
+        }
+    }
+
+    /// Ping to update the votes according to current stake of validators.
+    pub fn ping(&mut self) {
+        assert!(self.result.is_none(), "Voting has already ended");
+        let cur_epoch_height = env::epoch_height();
+        if cur_epoch_height != self.last_epoch_height {
+            for account_id in self.votes.keys().into_iter().collect::<Vec<_>>() {
+                let account_current_stake = env::validator_stake(&account_id);
+                let account_voted_stake = self.votes.remove(&account_id).unwrap();
```

Comment:
Self.votes.get?
After remove the votes will be gone?
Also why does this has a value? Should this just be a hashset of Validators that voted?

Reply:
Good point. I am retarded.

Reply:
Actually it's needed when a validator who already voted calls `vote` again. If we don't store their voted stake we don't know how to update `total_voted_stake`.